api/models/fn.go

```diff
@@ -0,0 +1,235 @@
+
+	if f.AppID == "" {
+		return ErrFnsMissingAppID
+	}
+
+	if f.Image == "" {
+		return ErrFnsMissingImage
+	}
+
+	switch f.Format {
+	case FormatDefault, FormatHTTP, FormatJSON, FormatCloudEvent:
+	default:
+		return ErrFnsInvalidFormat
+	}
+
+	if f.Timeout <= 0 || f.Timeout > MaxTimeout {
+		return ErrInvalidTimeout
+	}
+
+	if f.IdleTimeout <= 0 || f.IdleTimeout > MaxIdleTimeout {
+		return ErrInvalidIdleTimeout
+	}
+
+	if f.Memory < 1 || f.Memory > MaxMemory {
+		return ErrInvalidMemory
+	}
+
+	return f.Annotations.Validate()
+}
+
+func (f *Fn) Clone() *Fn {
+	clone := new(Fn)
+	*clone = *f // shallow copy
+
+	// now deep copy the maps
+	if f.Config != nil {
+		clone.Config = make(Config, len(f.Config))
+		for k, v := range f.Config {
+			clone.Config[k] = v
+		}
+	}
+	if f.Annotations != nil {
+		clone.Annotations = make(Annotations, len(f.Annotations))
+		for k, v := range f.Annotations {
+			// TODO technically, we need to deep copy the bytes
+			clone.Annotations[k] = v
+		}
+	}
+	return clone
+}
+
+func (f1 *Fn) Equals(f2 *Fn) bool {
+	// start off equal, check equivalence of each field.
+	// the RHS of && won't eval if eq==false so config/headers checking is lazy
+
+	eq := true
+	eq = eq && f1.AppID == f2.AppID
+	eq = eq && f1.Image == f2.Image
+	eq = eq && f1.Memory == f2.Memory
+	eq = eq && f1.CPUs == f2.CPUs
+	eq = eq && f1.Format == f2.Format
+	eq = eq && f1.Timeout == f2.Timeout
+	eq = eq && f1.IdleTimeout == f2.IdleTimeout
+	eq = eq && f1.Config.Equals(f2.Config)
+	eq = eq && f1.Annotations.Equals(f2.Annotations)
+	// NOTE: datastore tests are not very fun to write with timestamp checks,
+	// and these are not values the user may set so we kind of don't care.
+	//eq = eq && time.Time(f1.CreatedAt).Equal(time.Time(f2.CreatedAt))
+	//eq = eq && time.Time(f2.UpdatedAt).Equal(time.Time(f2.UpdatedAt))
+	return eq
+}
+
+// Update updates fields in f with non-zero field values from new, and sets
+// updated_at if any of the fields change. 0-length slice Header values, and
+// empty-string Config values trigger removal of map entry.
+func (f *Fn) Update(patch *Fn) {
+	original := f.Clone()
+
+	if patch.AppID != "" {
+		f.AppID = patch.AppID
+	}
+	if patch.Image != "" {
+		f.Image = patch.Image
+	}
+	if patch.Memory != 0 {
+		f.Memory = patch.Memory
+	}
+	if patch.CPUs != 0 {
+		f.CPUs = patch.CPUs
+	}
+	if patch.Timeout != 0 {
+		f.Timeout = patch.Timeout
+	}
+	if patch.IdleTimeout != 0 {
+		f.IdleTimeout = patch.IdleTimeout
+	}
+	if patch.Format != "" {
+		f.Format = patch.Format
+	}
+	if patch.Config != nil {
+		if f.Config == nil {
+			f.Config = make(Config)
+		}
+		for k, v := range patch.Config {
+			if v == "" {
+				delete(f.Config, k)
+			} else {
+				f.Config[k] = v
+			}
+		}
+	}
+
+	f.Annotations = f.Annotations.MergeChange(patch.Annotations)
+
+	if !f.Equals(original) {
+		f.UpdatedAt = strfmt.DateTime(time.Now())
+	}
+}
+
+type FnFilter struct {
+	AppID string // this is exact match
+	Image string // this is exact match
+
```

Comment:
Don't we need a few more things here? Name for example?

Reply:
lets add later as needed - I'd go so far as to stay `Image` is a feature unbacked by need.